### PR TITLE
Allow aliases and template aliases to be considered for replacement

### LINKED ISSUE #89: Replacing template alias that does not use its template parameter with one that does is broken
``` c++
// bindgen-flags: -- --std=c++14

namespace JS {
namespace detail {

// Notice how this doesn't use T.
template <typename T>
using MaybeWrapped = int;

}

template <typename T>
class Rooted {
    detail::MaybeWrapped<T> ptr;
};

}

/// But the replacement type does use T!
///
/// <div rustbindgen replaces="MaybeWrapped" />
template <typename T>
using replaces_MaybeWrapped = T;
```

Oh man this took a while to track down. I'm hitting this in the SM bindings because we don't understand `mozilla::Conditional` and therefore don't understand that it _is_ using `T` and we shouldn't filter the type parameter out for being unused, but we aren't smart enough to understand that.

I "fixed" this with this diff:

``` diff
             TypeKind::Alias(_, inner) => {
                 let parent_args = ctx.resolve_item(self.parent_id())
                    .applicable_template_args(ctx);
                 let inner = ctx.resolve_type(inner);
                 // Avoid unused type parameters, sigh.
                 parent_args.iter().cloned().filter(|arg| {
-                    let arg = ctx.resolve_type(*arg);
-                    arg.is_named() && inner.signature_contains_named_type(ctx, arg)
+                    // let arg = ctx.resolve_type(*arg);
+                    // arg.is_named() && inner.signature_contains_named_type(ctx, arg)
+
+                    true
                 }).collect()
             }
```

however that breaks perfectly reasonable things like `tests/headers/template_typedefs.h`.

Clearly we need to be smarter about type parameters, but I'm not 100% sure how to do that yet, and I'm still fumbling around this code base.

### PR DESCRIPTION
**Pull request by fitzgen**
Fixes #89.

I'm not exactly *happy* with the way this is implemented (making `real_canonical_name` public so that we can use it in replacement lookups) but I'm not sure of a better way without refactoring most of how naming works right now.

r? @emilio

### CODE REVIEW

File: src/ir/context.rs [hunk 1]
```diff
@@ -320,23 +320,28 @@ impl<'ctx> BindgenContext<'ctx> {
         let mut replacements = vec![];
 
         for (id, item) in self.items.iter() {
+            // Calls to `canonical_name` are expensive, so eagerly filter out
+            // items that cannot be replaced.
+
```
Comment (emilio) [inline src/ir/context.rs]: nit: this newline may be removed

File: src/ir/context.rs [hunk 2]
```diff
@@ -320,23 +320,28 @@ impl<'ctx> BindgenContext<'ctx> {
...
-            let ci = match ty.as_comp() {
-                Some(ci) => ci,
-                None => continue,
-            };
-
-            if ci.is_template_specialization() {
-                continue;
+            match *ty.kind() {
+                TypeKind::Comp(ref ci) if !ci.is_template_specialization() => {}
+                TypeKind::TemplateAlias(_, _) => {}
```
Comment (emilio) [inline src/ir/context.rs]: nit: Maybe `TypeKind::TemplateAlias(_, _) | TypeKind::Alias(_, _) => {}`?

**Comment by fitzgen** [comment]
Fixed nits!

**Comment by emilio** [comment]
@bors-servo r+

Thanks!

**Pull request merged**